### PR TITLE
New Protocol NFS

### PR DESCRIPTION
**Pull request by termanix**
Lately I've been thinking about what NetExec needs in development and I thought seeing NFS shares would also help with development.

It's using RPC and enumerating NFS Shares and if accessible, it finds files recursively.

Together with @Marshall-Hallenbeck, we were able to bring it to its current state (Thank you Marshall for helping). I hope it will be a useful protocol for everyone.

Here it is some screenshots:

My test lab,

192.168.37.130 : Server 2019 Domain Controller
192.168.37.131 : Ubuntu 6.8.0.-31 

image

image

And also including UID brute force for if anonymously authentication fail.

image

### CODE REVIEW

**Comment by mpgn** [comment]
Crazy !!! 🎉🎉🎉

**Comment by Zamanry** [comment]
Hell yeah! I’ve been wanting something like this for a while!!!

File: nxc/protocols/nfs/proto_args.py [hunk 1]
```diff
@@ -5,6 +5,7 @@ def proto_args(parser, parents):
     dgroup = ldap_parser.add_argument_group("NFS Mapping/Enumeration", "Options for Mapping/Enumerating NFS")
     dgroup.add_argument("--shares", action="store_true", help="Authenticate locally to each target")
     dgroup.add_argument("--shares-list", action="store_true", help="Listing enumerated shares")
+    dgroup.add_argument("--bruteforce-uid", nargs="?", type=int, const=4000, metavar="MAX_UID", help="Enumerate shares by bruteforcing UIDs")
```
Comment (mpgn) [inline nxc/protocols/nfs/proto_args.py:8]: to keep consistency with pro SMB with option (`--rid-brute`), maybe it should be rename `--uid-brute`

Comment (termanix) [reply nxc/protocols/nfs/proto_args.py:8]: --uid-brute sounds good! I will rename it. Also yes, it seems like SMB

**Review by termanix — COMMENTED** [review]
to keep consistency with SMB's ```--rid-brute```

**Comment by NeffIsBack** [comment]
Improvements & Fixes:
- Moved source for PyNfsClient to the Pennyw0rth organisation
- Fixed RPC authentication in PyNfsClient
  - Added detection for kerberos requirement
  - Fixed socket errors when authentication fails
  - Merged ipv6 fix from an upstream PR
- Fixed recursion depth when enumerating shares
- Other code improvements

ToDos:
- Fix now spammed output when RID-Bruteforcing
- Fix ugly output

Example screenshot:
image

**Comment by termanix** [comment]
> Improvements & Fixes:
> 
> * Moved source for PyNfsClient to the Pennyw0rth organisation
> * Fixed RPC authentication in PyNfsClient
>   
>   * Added detection for kerberos requirement
>   * Fixed socket errors when authentication fails
>   * Merged ipv6 fix from an upstream PR
> * Fixed recursion depth when enumerating shares
> * Other code improvements
> 
> ToDos:
> 
> * Fix now spammed output when RID-Bruteforcing
> * Fix ugly output
> 
> Example screenshot: image

Now it belongs to netexec format, thank you Neff. I wil work on to dos on my free.

**Comment by NeffIsBack** [comment]
Added ipv6 support to the NfsClient library as well as the protocol:
image

**Comment by termanix** [comment]
All flags and output works nice /w Neff

image

**Comment by NeffIsBack** [comment]
Actually we can query the uid needed for the share lol. Implemented an autodetecting feature that will automatically set the uid matching to the listed repository:
image

**Comment by termanix** [comment]
Download and upload files added.
image

To Do: 
    When uploaded a file, we cant read it. 
image

**Comment by NeffIsBack** [comment]
`--get-file` and `--put-file` are working as well 🚀 
Escept for windows
image

**Comment by termanix** [comment]
To Do: Only left puf-file bug fixes for windows and linux.

**Comment by NeffIsBack** [comment]
Up&Download are working now against Linux&Windows. Also found the solution for setting file permissions 🎉 Default is 777 for now
image
image
image

**Comment by NeffIsBack** [comment]
@termanix if you can verify the behaviour on your side we should be good to go 🚀

**Comment by termanix** [comment]
@NeffIsBack On my side everything is fine now! It can be merge after your last review. 🚀

**Comment by NeffIsBack** [comment]
LGTM:
image

**Review by NeffIsBack — APPROVED** [review]
Everything tested, we should be good 🚀

**Review by mpgn — APPROVED** [review]

**Comment by mpgn** [comment]
Excellent work from both of you @NeffIsBack @termanix ! 🎉

**Pull request merged**